src/debugger/Debugger.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import type { BabelNode } from "babel-types";
+import { Breakpoint } from "./Breakpoint.js";
+import { ExecutionContext } from "../realm.js";
+import invariant from "../invariant.js";
+const fs = require("fs");
+
+export class Debugger {
+  constructor(dbgFileLines: Array<string>) {
+    this.breakpoints = new Map();
+    this.prevBreakLine = 0;
+    this.prevBreakCol = 0;
+    this.parseCommands(dbgFileLines);
+  }
+  breakpoints: Map<number, Breakpoint>;
+  prevBreakLine: number;
+  prevBreakCol: number;
+
+  serializeContext(context: ExecutionContext) {
+    return context.loc;
+  }
+
+  checkForActions(ast: BabelNode, contextStack: Array<ExecutionContext>) {
+    this.checkForBreakpoint(ast, contextStack);
+  }
+
+  proceedBreakpoint(lineNum: number, colNum: number): boolean {
+    if (this.breakpoints.has(lineNum) && this.breakpoints.get(lineNum).enabled) {
+      if (lineNum === this.prevBreakLine) {
+        if (colNum === this.prevBreakCol) {
+          this.prevBreakCol = 0;
+        } else {
+          this.prevBreakCol = colNum;
+        }
+        return false;
+      }
+      this.prevBreakLine = lineNum;
+      return true;
+    }
+    return false;
+  }
+
+  checkForBreakpoint(ast: BabelNode, contextStack: Array<ExecutionContext>) {
+    invariant(ast.loc);
+    let location = ast.loc;
+    if (location.start.line === location.end.line) {
+      let lineNum = location.start.line;
+      let colNum = location.start.column;
+      if (!this.proceedBreakpoint(lineNum, colNum)) return;
+
+      console.log("Stopped for breakpoint on line " + lineNum);
+      this.sendDebugInfo(ast, contextStack, lineNum);
+
+      let lastPoll = Date.now();
+      let blocking = true;
+      let contents = "";
+      while (blocking) {
+        if (Date.now() - lastPoll > 1000) {
+          contents = fs.readFileSync("./src/debugger/.sessionlogs/proxy2debugger.txt", "utf8").toString().split("\n");
+          if (contents[0] === "proceed " + lineNum) {
+            blocking = false;
+          } else {
+            this.parseCommands(contents);
+          }
+          lastPoll = Date.now();
+        }
+      }
+      fs.writeFileSync("./src/debugger/.sessionlogs/proxy2debugger.txt", "");
+    }
+  }
+
+  sendDebugInfo(ast: BabelNode, contextStack: Array<ExecutionContext>, lineNum: number) {
+    fs.writeFileSync("./src/debugger/.sessionlogs/debugger2proxy.txt", "breakpoint " + lineNum);
+    // for (let i = 0; i < contextStack.length; i++) {
```

Comment:
It is never OK to check in code inside comments. If this pull request is only intended for early feed back, make that clear in the title of the pull request.

Reply:
Removed